types/src/build_version/mod.rs

```diff
@@ -0,0 +1,143 @@
+const MINOR_ONLY_MASK: u32 = 0x00ff00;
+const PATCH_ONLY_MASK: u32 = 0x0000ff;
+
+//api versions
+const FULL_API_VERSION: Version = new_version(0, 17, 0);
+const MINER_API_VERSION: Version = new_version(0, 15, 0);
+const WORKER_API_VERSION: Version = new_version(0, 15, 0);
+
+/// a constant for the block delay process in seconds
+pub const BLOCK_DELAY_SECS: u64 = 4;
+
+lazy_static! {
+    static ref CURRENT_COMMIT: String = {
+        let output = Command::new("git")
+            .args(&[
+                "describe",
+                "--always",
+                "--match=NeVeRmAtch",
+                "--dirty",
+                "2>/dev/null",
+            ])
+            .output()
+            .map(|s| s.stdout)
+            .unwrap_or_else(|_| {
+                Command::new("git")
+                    .args(&["rev-parse", "--short", "HEAD", "2>/dev/null"])
+                    .output()
+                    .map(|s| s.stdout)
+                    .unwrap_or_default()
+            });
+        String::from(std::str::from_utf8(&output).unwrap_or_default())
+    };
+    pub static ref BUILD_TYPE: RwLock<BuildType> = RwLock::new(BuildType::BuildDefault);
+    pub static ref RUNNING_NODE_TYPE: RwLock<NodeType> = RwLock::new(NodeType::Full);
+}
+
+/// represents the current version of the api
+
+#[derive(Serialize)]
+#[serde(rename_all = "PascalCase")]
+pub struct APIVersion {
+    pub version: String,
+    pub api_version: Version,
+    pub block_delay: u64,
+}
+
+/// integer based value on version information. Highest order bits for Major, Mid order for Minor and lowest for Patch
+#[derive(Serialize)]
+pub struct Version(u32);
+
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum BuildType {
+    BuildDefault = 0x0,
+    Build2k = 0x1,
+    BuildDebug = 0x2,
+}
+
+/// the type of node that is running
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum NodeType {
+    Unknown = 0,
+    Full = 1,
+    Miner = 2,
+    Worker = 3,
+}
+
+impl Display for NodeType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "{}", self)
+    }
+}
+
+impl BuildType {
+    fn to_str(&self) -> &str {
+        match self {
+            BuildType::BuildDefault => "",
+            BuildType::Build2k => "+debug",
+            BuildType::BuildDebug => "+2k",
+        }
+    }
+}
+
+const fn new_version(major: u32, minor: u32, patch: u32) -> Version {
+    Version(major << 16 | minor << 8 | patch)
+}
+
+/// gets current user version
+pub async fn user_version() -> String {
+    BUILD_VERSION.to_owned() + &*BUILD_TYPE.read().await.to_str() + &CURRENT_COMMIT
+}
+impl Version {
+    fn ints(&self) -> (u32, u32, u32) {
+        let v = self.0;
+        (
+            (v & MAJOR_ONLY_MASK) >> 16,
+            (v & MINOR_ONLY_MASK) >> 8,
+            v & PATCH_ONLY_MASK,
+        )
+    }
+}
+
+impl PartialEq for Version {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 & MINOR_MASK == other.0 & MINOR_MASK
+    }
+}
+
+impl ToString for Version {
```

Comment:
why not display?

Reply:
matching lotus for concatenation but can change!